### PR TITLE
improve defmt version mismatch diagnostics

### PR DESCRIPTION
**Pull request by japaric**
offer a different suggestion based on whether a git or crates.io version is being used

should be merged after #244

### CODE REVIEW

File: decoder/src/lib.rs
Comment (japaric) [inline decoder/src/lib.rs:139]: `panic-probe` currently lives in the `probe-run` repo so it's not really possible to pin to the correct defmt rev (`panic-probe` also already pins to a particular `defmt` and that cannot be changed from an application project). I think that to simplify the use of the git-version of defmt we'll have to move panic-probe (back) into the defmt repo. This PR is blocked on that migration.

Comment (japaric) [reply decoder/src/lib.rs:139]: migration PR: #244

File: decoder/src/lib.rs
Comment (Lotterleben) [inline decoder/src/lib.rs:108]: question: why are these defined within the function instead of outside?

Comment (japaric) [reply decoder/src/lib.rs:108]: the `enum` is only used within the function body so putting it inside avoids adding yet another type to the outer (`src/lib.rs`) type namespace.

Another way to keep the namespace tidy is use modules so this could have been written as: `mod version { enum Kind { .. } pub fn check(version: &str) -> Result<(), String> { .. } }`. We have not used modules in this crate at all so the file is over 1.6K lines ... we should eventually come back and refactor this into modules.

Comment (Lotterleben) [reply decoder/src/lib.rs:108]: alrighty, thank you for the clarification! +1 to the refactoring, but that's for another day ofc

**Comment by japaric** [comment]
reminder to self: publish a new patch version of `defmt-decoder` after this is merged

**Review by Lotterleben — APPROVED** [review]

**Pull request merged**
